### PR TITLE
toggle hooks in middleware

### PR DESCRIPTION
**Pull request by robinfehr**
- makes it optional if the hooks are piped to the middleware.
we might wan't to fine grain it so we can choose which hooks.
the default is 'included' to not break current implementations.
```ts
addMiddleware(node, middleware: (actionDescription, next) => any, includeHooks = true)
```

- throws on omitting next() within a middleware

### CODE REVIEW

File: packages/mobx-state-tree/src/core/action.ts [hunk 1]
```diff
@@ -126,41 +135,61 @@ export function decorate<T extends Function>(middleware: IMiddlewareHandler, fn:
...
-        const handler = handlers[index++]
-        if (handler) return handler(call, runNextMiddleware)
-        else return mobxAction(originalFn).apply(null, baseCall.args)
+        const middleware = middlewares[index++]
+        const handler = middleware && middleware.handler
+        const invokeHandler = () => {
+            const next = handler(call, runNextMiddleware)
+            if (!next && index < middlewares.length && process.env.NODE_ENV !== "production") {
+                const node = getStateTreeNode(call.tree)
+                fail(
```
Comment (mweststrate) [inline packages/mobx-state-tree/src/core/action.ts]: I'm not sure about this check, there are cases where it is fine that `next()` isn't called, for example if you have middleware that prevents modifications, or only applies it to a copy, etc. I had a few cases before. I am aware it is pretty tricky if you forget it in your middleware, but I think that is an inherent risk about writing your own middleware and hopefully docs help preventing mistakes there :)

File: API.md [hunk 2]
```diff
@@ -124,7 +124,9 @@ For more details, see the [middleware docs](docs/middleware.md)
 **Parameters**
 
 -   `target` **IStateTreeNode** 
--   `middleware`  
+-   `middleware`  **IMiddleware**
+-   `includeHooks` **([boolean](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean) | any)** indicates whether the hooks should be piped to the middleware.
```
Comment (mweststrate) [inline API.md]: elobarote; `that is, should the middleware be run for built in methods like `applySnapshot`

File: packages/mobx-state-tree/src/core/action.ts (hunk 1 above)
Comment (robinfehr) [reply packages/mobx-state-tree/src/core/action.ts]: fair point about aborting the middleware queue.
what about returning false if it the abort was intended and only throw on undefined?

Comment (robinfehr) [reply packages/mobx-state-tree/src/core/action.ts]: e.g. an `abort()` next to `next()`

**Comment by mweststrate** [comment]
Merged!

**Pull request merged**